src/AutoRest.CSharp/LowLevel/AutoRest/DpgOutputLibrary.cs

```diff
@@ -10,79 +10,181 @@
-            ClientOptions = clientOptions;
-        }
 
-        private AspDotNetExtensionTypeProvider? _aspDotNetExtension;
-        public AspDotNetExtensionTypeProvider AspDotNetExtension => _aspDotNetExtension ??= new AspDotNetExtensionTypeProvider(RestClients, Configuration.Namespace, _sourceInputModel);
+            var defaultNamespace = Configuration.Namespace;
+            var libraryName = Configuration.LibraryName;
+            var defaultModelNamespace = TypeProvider.GetDefaultModelNamespace(null, defaultNamespace);
+
+            _enums = CreateEnums(rootNamespace.Enums, defaultModelNamespace, TypeFactory, sourceInputModel);
+            _models = CreateModels(rootNamespace.Models, defaultModelNamespace, TypeFactory, sourceInputModel);
 
-        private ModelFactoryTypeProvider? _modelFactoryProvider;
-        public ModelFactoryTypeProvider? ModelFactory => _modelFactoryProvider ??= ModelFactoryTypeProvider.TryCreate(AllModels, _sourceInputModel);
+            var allModels = new List<TypeProvider>(_enums.Values);
+            allModels.AddRange(_models.Values);
+            _privateAllModels = allModels;
+
+            AllModels = isTspInput ? allModels : Array.Empty<TypeProvider>();
+            RestClients = CreateClients(topLevelClientInfos, clientOptions, rootNamespace, TypeFactory, libraryName, sourceInputModel);
+            ClientOptions = clientOptions;
+            ModelFactory = ModelFactoryTypeProvider.TryCreate(AllModels, sourceInputModel);
+            AspDotNetExtension = new AspDotNetExtensionTypeProvider(RestClients, Configuration.Namespace, sourceInputModel);
+        }
 
         public override CSharpType ResolveEnum(InputEnumType enumType)
         {
-            if (!_isTspInput || enumType.Usage == InputModelTypeUsage.None)
+            if (_enums.TryGetValue(enumType with {IsNullable = false}, out var typeProvider))
             {
-                return TypeFactory.CreateType(enumType.EnumValueType);
-            }
-
-            if (_enums.TryGetValue(enumType, out var typeProvider))
-            {
-                return typeProvider.Type;
+                return typeProvider.Type.WithNullable(enumType.IsNullable);
             }
 
             throw new InvalidOperationException($"No {nameof(EnumType)} has been created for `{enumType.Name}` {nameof(InputEnumType)}.");
         }
 
         public override CSharpType ResolveModel(InputModelType model)
-            => _models.TryGetValue(model, out var modelTypeProvider) ? modelTypeProvider.Type : new CSharpType(typeof(object), model.IsNullable);
+            => _models.TryGetValue(model with {IsNullable = false}, out var modelTypeProvider) ? modelTypeProvider.Type.WithNullable(model.IsNullable) : new CSharpType(typeof(object), model.IsNullable);
+
+        public override CSharpType? FindTypeByName(string originalName) => _privateAllModels.Where(m => m.Declaration.Name == originalName)?.Select(m => m.Type).FirstOrDefault();
```

Comment:
would it be more efficiant if we have
```suggestion
        public override CSharpType? FindTypeByName(string originalName) => _privateAllModels.FirstOrDefault(m => m.Declaration.Name == originalName)?.Type;
```